text/0002-new-architecture.md

```diff
@@ -0,0 +1,243 @@
+processing pipeline really could be separated entirely from the rest of the System.  However the same is true for quite
+a lot of different parts of the product.  For instance certain parts of the user experience are sufficiently idependent
+in the UI code today already that they could turn into a separate service that just supplies UI views together with some
+APIs.  While it's unclear if the existing experiences are worth splitting up into new services, it's probably likely that
+we can at least structure their code so that we can reduce the total amount of pieces of code that need to be checked and
+tested against changes.
+
+In the ideal situation a change to the settings page of a project for instance does not need to run the UI tests for
+performance views etc. unless the settings are in fact related to that component.
+
+### Service Declarations
+
+The biggest limiting factor today in creating new services is the configuration of these services.  When an engineer adds
+a new service there is typically a followup PR against the internal ops repo to configure these services for the main
+sentry installation.  Then there is some sort of ad-hoc workaround to make the whole thing work in local development, how
+to make it work on a single tenant or self hosted installation.  Each new services undergoes a new phase of rediscovery how
+suboptimal the rules and policies around scaling these services are.
+
+Ideally we were able to describe the services in a config file (YAML?) which then can be consumed by deploy tools to
+correctly provision and auto scale services.  This could either be a Sentry specific format or we adopt something that
+has already been tested.
+
+### Serverless Function Declarations for Consumers
+
+The smallest unit of execution at Sentry is not a docker image but in fact a function.  We currently have no real serverless
+setup but we have a few functions that are run as a kafka consumer, queue worker or similar.  Like services making these
+scale is quite tricky and in some situations we need to needlessly spawn more processes and containers even though they
+could be colocated.
+
+## Data Store
+
+These are abstract changes to the data store.  Largely this is not explored yet but some things are known.
+
+### High Cardinality Metrics
+
+As we are extracting metrics data from the existing transaction system we have effectively unlimited cardinality
+in the data stream coming in.  The system currently applies various different attempts of dealing with this problem
+where a lot of this is done on the Relay side.  This comes from the combination of the hard cardinality limited in
+Relay but also by attempting to not generate high cardinality data in the clients.
+
+However it's likely that we will be unable to reduce cardinality in the long run and the data model should ideally
+be able to represent this high cardinality data.
```

Comment:
Is supporting high cardinality metrics the first problem to solve ?
The product fundamentally relies on the assumption that real world data is low cardinality (at least for what concerns transaction names). The product concept falls apart if that is not true, i.e. if you cannot group by transaction names effectively. This is because the whole product experience is based on aggregating by transaction names.

So, knowing that real world data is high cardinality, shouldn't we first decide whether we change the product assumption (and then build a high cardinality storage) or whether we keep the product assumption and fix data cardinality instead ?

Reply:
@fpacifici I think you're mostly thinking about transaction name, but people are also wishing for significantly more tags on performance metrics that the product would be able to deal with very well... but not the infrastructure.

I think it would help to note down some examples here of concrete product features or queries that we would want to enable with this.

Reply:
Some examples of high cardinality data outside of URLs:

* User / Customer ID
* Shopping Cart ID
* Request ID
* IP address